### PR TITLE
kex2 security upgrade

### PR DESCRIPTION
**Pull request by maxtaco**
- don't send PPS in Hello(); rather, send the provisonee's encryption public key over in the reply to Hello, and send the encryption of PPS over with DidCounterSign()
- test for v1 v v2 compatibility and fallback

### CODE REVIEW

**Comment by maxtaco** [comment]
Hand-tested all the scenarios:

1. new provisioner, new provisionee
1. old provisioner, new provisionee
1. new provisioner, old provisionee

File: go/kex2/provisioner.go [hunk 1]
```diff
@@ -162,6 +173,47 @@ func (p *provisioner) runProtocolWithCancel() (err error) {
...
+func (p *provisioner) runProtocolV2() (fallback bool, err error) {
+	cli := keybase1.Kex2Provisionee2Client{Cli: rpc.NewClient(p.xp, nil)}
+	var helloArg keybase1.Hello2Arg
+	helloArg, err = p.arg.Provisioner.GetHello2Arg()
+	if err != nil {
+		return false, err
+	}
+	var res keybase1.Hello2Res
+	if res, err = cli.Hello2(context.TODO(), helloArg); err != nil {
+		if strings.Contains(err.Error(), "protocol not found: keybase.1.Kex2Provisionee2") {
```
Comment (maxtaco) [inline go/kex2/provisioner.go]: this is slightly ugly, but not much we can do subject to backwards compatibility here.

File: go/engine/kex2_provisioner.go [hunk 2]
```diff
@@ -210,6 +226,27 @@ func (e *Kex2Provisioner) CounterSign(input keybase1.HelloRes) (sig []byte, err
...
+	key, err = libkb.ImportKeypairFromKID(input.EncryptionKey)
+	if err != nil {
+		return output, err
+	}
+	output.Sig, err = e.CounterSign(input.SigPayload)
+	if err != nil {
+		return output, err
+	}
+	var ppsPacked []byte
+	ppsPacked, err = libkb.MsgpackEncode(e.pps)
```
Comment (patrickxb) [inline go/engine/kex2_provisioner.go]: What's the purpose of msgpack encoding pps before encrypting?

Comment (maxtaco) [inline go/engine/kex2_provisioner.go]: pps has 2 fields -- the stream itself and the generation number.

**Comment by patrickxb** [comment]
LGTM

**Pull request merged**